chips/virtio/src/queues/split_queue.rs

```diff
@@ -0,0 +1,911 @@
+    }
+
+    /// Whether the available ring of the Virtqueue is full.
+    pub fn is_full(&self) -> bool {
+        !self.empty.get() && self.start.get() == self.end.get()
+    }
+
+    /// Try to insert an element into the Virtqueue available ring.
+    ///
+    /// If there is space in the Virtqueue's available ring, this increments the
+    /// internal state and returns the index of the element to be
+    /// written. Otherwise, it returns `None`.
+    pub fn insert(&self) -> Option<u16> {
+        if !self.is_full() {
+            let pos = self.end.get();
+            self.end.set(self.ring_wrapping_add(pos, 1));
+            self.empty.set(false);
+            Some(pos)
+        } else {
+            None
+        }
+    }
+
+    /// Try to remove an element from the Virtqueue available ring.
+    ///
+    /// If there is an element in the Virtqueue's available ring, this removes
+    /// it from its internal state and returns the index of that element.
+    pub fn pop(&self) -> Option<u16> {
+        if !self.is_empty() {
+            let pos = self.start.get();
+            self.start.set(self.ring_wrapping_add(pos, 1));
+            if self.start.get() == self.end.get() {
+                self.empty.set(true);
+            }
+            Some(pos)
+        } else {
+            None
+        }
+    }
+}
+
+/// Internal representation of a slice of memory passed held in the Virtqueue.
+///
+/// Because of Tock's architecture combined with Rust's reference lifetime
+/// rules, buffers are generally passed around as `&mut [u8]` slices with a
+/// `'static` lifetime. Thus, clients pass a mutable static reference into the
+/// Virtqueue, losing access. When the device has processed the provided buffer,
+/// access is restored by passing the slice back to the client.
+///
+/// However, clients may not wish to expose the full slice length to the
+/// device. They cannot simply subslice the slice, as that would mean that
+/// clients loose access to the "sliced off" portion of the slice. Instead,
+/// clients pass a seperate `length` parameter when inserting a buffer into a
+/// virtqueue, by means of the [`VirtqueueBuffer`] struct. This information is
+/// then written to the VirtIO descriptor.
+///
+/// Yet, to be able to reconstruct the entire buffer and hand it back to the
+/// client, information such as the original length must be recorded. We cannot
+/// retain the `&'static mut [u8]` in scope though, as the VirtIO device (host)
+/// writing to it would violate Rust's memory aliasing rules. Thus, we convert a
+/// slice into this struct (converting the slice into its raw parts) for
+/// safekeeping, until we reconstruct a byte-slice from it to hand it back to
+/// the client.
+///
+/// While we technically retain an identical pointer in the
+/// [`VirtqueueDescriptors`] descriptor table, we record it here nonetheless, as
+/// a method to sanity check internal buffer management consistency.
+struct SharedDescriptorBuffer<'b> {
```

Comment:
I think this is fine here for now, but it seems like this might eventually be useful as a general Tock type for buffers passed to DMA etc, and could go into `kernel/utils`

Reply:
Yes, that's right. This is added explicitly here as #2637 hasn't progressed to an actual PR yet, but it's pretty high up my todo list and will essentially be a fancier version of this struct, maintained in the kernel.

Reply:
Great! Try to make a note on #2367 to remove this once a fancier version is included